Riese/web user deactivation es

MartinRiese opened the pull request:
## Product Description

Since web uses can be deactivated per domain now (still behind a feature flag DEACTIVATE_WEB_USERS) we added an is_active flag to the domain memberships in ES model.

Previously there was only an is_active flag on the users themselves. For mobile users that meant they were active/deactivated on their domain. For web users it is used to mark if the SSO says they are active it they use one.

This PR also changes all ES queries that filter users out by activity. Since is_active=True was a default filter, this is almost all of them. They now have to provide a domain to check if the user is active on that particular domain. The behavior for mobile user should stay unchanged.

Prominent places where ES user queries are used:
* User export
* Report filters (those have some wonkiness to them to begin with and we tried our best not to make any changes to the behavior. But might add some fixes to address inconsistency later)
* Login as

## Technical Summary

Epic: https://dimagi.atlassian.net/browse/USH-6032
* https://dimagi.atlassian.net/browse/USH-6005
* https://dimagi.atlassian.net/browse/USH-6000
* https://dimagi.atlassian.net/browse/USH-6040

## Feature Flag

DEACTIVATE_WEB_USERS for deactivating web uses. But these changes will be active without the FF too.

## Safety Assurance

### Safety story

Testing by devs and QA

### Automated test coverage

Had to update quite s few test and added more for good measure.

### QA Plan

https://dimagi.atlassian.net/browse/QA-7879

### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

esoergel committed:
Make domain filter handle is_active status

MartinRiese committed:
Fix usage of UserES

Move include inactive into domain call

MartinRiese committed:
Fix usage of UserES

Add include_inactive to active_user_es_query so it can pass it to domain

MartinRiese committed:
Fix UserES usage

Straight forward replacement

MartinRiese committed:
Fix UserES usage

is_active is covered by domain now. By default it only includes
active users

esoergel committed:
Remove default activity filters

MartinRiese committed:
Fix UserES usage

MartinRiese committed:
Fix UserES usage: User active on any domain

For mass emails we just care that the user is
active on at least one domain

MartinRiese committed:
Fix lint errors

MartinRiese committed:
Fix tests

MartinRiese committed:
Fix UserES usage

Domain now filters out inactive users by default
Add right flags to get intended users.

MartinRiese committed:
Fix UserES usage

Domain now filters out inactive users by default
Add right flags to get intended users.

esoergel committed:
A few easy fixes

esoergel committed:
Make domain available to locations dbaccessors

MartinRiese committed:
Fix lint error

esoergel committed:
That's the last of 'em

esoergel committed:
Update web users list

esoergel committed:
Remove duplicate domains fn in favor of domain

MartinRiese committed:
Fix test

functions calls were changed in a previous commit

MartinRiese committed:
Fix test: domain should be passed in the builder

MartinRiese committed:
Fix expand mobile worker filter

Fix its usage of the UserES filter to match the new
functionality. Update the test data

MartinRiese committed:
Remove debug statement

MartinRiese committed:
Add remove_default_filters since it removes more than just is_active

MartinRiese committed:
Revert "Revert changes in enterprise, reports, saved_reports"

This reverts commit 8b728c6e16f167d099d45d8b6402953e10714511.

MartinRiese committed:
Revert "Revert some non essential changes to be addressed in a later PR"

This reverts commit bb03b3761c76ed5056354a7619a240b25690e0ec.

MartinRiese committed:
Merge branch 'master' into riese/web-user-deactivation-es

esoergel committed:
This is now implied by the `domain` filter

esoergel committed:
Move updated "is active" logic into domain filter

When dealing with multiple domains, NOT(is_inactive(...)) isn't correct,
as a user could be inactive on one domain, but active on another.  The
set of matching users between is_active and is_inactive can overlap

esoergel committed:
Update remaining references to `is_active`

MartinRiese committed:
Refactor base query to just domains

That way they can be reused in the other domain filters.

Also extracted allow_enterprise

esoergel committed:
Simplify TestEMWFilterOutput

esoergel committed:
In-line query for MW page

esoergel committed:
in-line get_search_users_in_domain_es_query again

esoergel committed:
remove last usages of show_inactive

MartinRiese committed:
Fix user report filter to apply location correctly

The location filter needs the be in the same clause as the activity
check.

Also, allow to search for active and inactive web users at the same
time.

esoergel committed:
Clarify filter selections

esoergel committed:
Add tests for location-restricted users

MartinRiese committed:
Create domain used in test.

MartinRiese committed:
Lint

esoergel committed:
Update UserESFake.domain to mirror UserES.domain

MartinRiese committed:
Merge branch 'master' into riese/web-user-deactivation-es

esoergel committed:
Support querying active users and specific

esoergel commented on corehq/apps/reports/filters/users.py:
```diff
@@ -365,46 +371,43 @@ def user_es_query(cls, domain, mobile_user_and_group_slugs, request_user):
-        elif HQUserType.DEACTIVATED in user_types:
-            q = q.show_only_inactive()
-
+        elif HQUserType.ACTIVE in user_types or HQUserType.DEACTIVATED in user_types:
```
just pushed that change, but feel free to rebut if I'm missing something.

MartinRiese replied:
You are right.

esoergel committed:
Fix filter for deactivated web users

esoergel committed:
Simplify locations access check

This is a deliberate behavior change as well, as there were paths
through this function that incorrectly circumvented location access
controls

minhaminha reviewed (approved):
Rushed the review a little bit but combined with past context and QA passing everything, I think it looks good
